pgx/src/array.rs

```diff
@@ -0,0 +1,215 @@
+use crate::datum::{Array, FromDatum};
+use crate::pg_sys;
+use core::ptr::{slice_from_raw_parts_mut, NonNull};
+use pgx_pg_sys::*;
+
+extern "C" {
+    /// # Safety
+    /// Does a field access, but doesn't deref out of bounds of ArrayType
+    fn pgx_ARR_DATA_PTR(arrayType: *mut ArrayType) -> *mut u8;
+    /// # Safety
+    /// Does a field access, but doesn't deref out of bounds of ArrayType
+    fn pgx_ARR_DIMS(arrayType: *mut ArrayType) -> *mut libc::c_int;
+    /// # Safety
+    /// Must only be used on a "valid" (Postgres-constructed) ArrayType
+    fn pgx_ARR_NELEMS(arrayType: *mut ArrayType) -> i32;
+    /// # Safety
+    /// Does a field access, but doesn't deref out of bounds of ArrayType
+    fn pgx_ARR_NULLBITMAP(arrayType: *mut ArrayType) -> *mut bits8;
+}
+
+/// Handle describing a bare, "untyped" pointer to an array,
+/// offering safe accessors to the various fields of one.
+#[repr(transparent)]
```

Comment:
👀 Had to look up `repr(transparent)`

Reply:
Hm. I'm not sure if it actually should be, tbh. Might want to look up the `len` and cache it on construction.

Reply:
> Hm. I'm not sure if it actually should be, tbh. Might want to look up the `len` and cache it on construction.

Ah, I believe I will learn what we should do when I construct the followups to this that wrap RawArray further to allow more "fully safe" handling of it. It may require changing some design specs given here.